includes/spam.php

```diff
@@ -0,0 +1,139 @@
+}
+if ( ! defined( 'PMPRO_SPAM_ACTION_TIME_LIMIT' ) ) {
+    define( 'PMPRO_SPAM_ACTION_TIME_LIMIT', 900 );  // in seconds
+}
+
+/**
+ * Is the current visitor a spammer?
+ * @since 2.7
+ */
+function pmpro_is_spammer() {
+    $is_spammer = false;
+    
+    $activity = pmpro_get_spam_activity();
+    if ( count( $activity ) >= PMPRO_SPAM_ACTION_NUM_LIMIT ) {
+        $is_spammer = true;
+    }
+    
+    return apply_filters( 'pmpro_is_spammer', $is_spammer, $activity );
+}
+
+/**
+ * Get spam activity by IP.
+ * @since 2.7
+ * @param string $ip    The IP address to get activity for.
+ * @return array|False  The array of activity if successful, false if no IP. 
+ */
+function pmpro_get_spam_activity( $ip = null ) {
+    if ( empty( $ip ) ) {
+        $ip = pmpro_get_ip();        
+    }
+    
+    // If we can't determine the IP, let's bail.
+    if ( empty( $ip ) ) {
+        return false;
+    }
+    
+    $transient_key = 'pmpro_spam_activity_' . $ip;
+    $activity = get_transient( $transient_key );    
+    if ( empty( $activity ) ) { $activity = []; }
+    
+    // Remove old items.
+	$new_activity = [];
+	$now = current_time( 'timestamp', true ); // UTC
+    foreach( $activity as $item ) {
+		if ( $item > $now-( PMPRO_SPAM_ACTION_TIME_LIMIT ) ) {
+			$new_activity[] = $item;
+		}
+	}
+    
+    return $new_activity;    
+}
+
+/**
+ * Track spam activity.
+ * When we hit a certain number, the spam flag will trigger.
+ * For now we are only tracking credit card declines their timestamps.
+ * IP address isn't a perfect way to track this, but it's the best we have.
+ *
+ * @since 2.7
+ * @param string $ip    The IP address to track activity for.
+ * @return bool         True if successful. False if we couldn't determine the IP.
+ */
+function pmpro_track_spam_activity( $ip = null ) {
+    if ( empty( $ip ) ) {
+        $ip = pmpro_get_ip();        
+    }
+    
+    // If we can't determine the IP, let's bail.
+    if ( empty( $ip ) ) {
+        return false;
+    }
+        
+    $activity = pmpro_get_spam_activity( $ip );
+    $now = current_time( 'timestamp', true ); // UTC
+    array_unshift( $activity, $now );
+    
+    // If we have more than the limit, don't bother storing them.
+    if ( count( $activity ) > PMPRO_SPAM_ACTION_NUM_LIMIT ) {
+        rsort( $activity );
+        $activity = array_slice( $activity, 0, PMPRO_SPAM_ACTION_NUM_LIMIT );
+    }
+    
+    // Save to transient.
+    $transient_key = 'pmpro_spam_activity_' . $ip;
+    set_transient( $transient_key, $activity );
+    
+    return true;
+}
+
+/**
+ * Clears all stored spam activity for an IP address.
+ * Note that the pmpro_get_spam_activity function clears out old values
+ * automatically, and this should only be used to completely clear the activity.
+ *
+ * @since 2.7
+ * @param string $ip    The IP address to clear activity for.
+ * @return bool         True if successful. False if we couldn't determine the IP.

```

Comment:
```suggestion
 * @since 2.7
 *
 * @param string|null $ip The IP address to clear activity for, or leave as null to attempt to determine current IP address.
 *
 * @return bool True if the clearing of activity was successful, or false if IP could not be determined.
 */
```